release-calendar/src/client/api-service.ts

```diff
@@ -23,24 +23,15 @@ export class ApiService implements ApiServiceInterface {
   private getPromotionRequest(url: string): Promise<Promotion[]> {
     return fetch(url).then((result) => result.json());
   }
-  private getPromotionRequests(url: string): Promise<Promotion[][]> {
-    return fetch(url).then((result) => result.json());
-  }
 
   async getReleases(): Promise<EventInput[]> {
-    const gridOfPromotions = await this.getPromotionRequests(SERVER_URL);
-    const collect: EventInput[] = [];
-    //each row is all the promotions that have occured in a single channel
-    gridOfPromotions.forEach((rowOfPromotions: Promotion[]) => {
-      const endDate: Date = new Date();
-      collect.push(new EventInput(rowOfPromotions[0], endDate));
-      for (let i = 1; i < rowOfPromotions.length; i++) {
-        collect.push(
-          new EventInput(rowOfPromotions[i], rowOfPromotions[i - 1].date),
-        );
-      }
+    const allPromotions = await this.getPromotionRequest(SERVER_URL);
+    const map = new Map<Channel, Date>();
+    return allPromotions.map((promotion: Promotion) => {
+      const date = map.get(promotion.channel) || new Date();
+      map.set(promotion.channel, promotion.date);
+      return new EventInput(promotion, date);
```

Comment:
I'm not sure I understand the two branches here - what happens in each case?

Reply:
Either the date is taken from the map, which contains the saved date from the previous promotion in that channel, or the date will be set to the current time because the promotion is the first promotion iterated over in its channel.

Reply:
Ah, I get it now. Makes sense :)